scapy/contrib/isotp.py

```diff
@@ -57,100 +52,1171 @@ def answers(self, other):
+
+            return True
+
+        def try_feed(self, identifier, ea, data):
+            first_byte = six.indexbytes(data, 0)
+            if len(data) > 1 and first_byte & 0xf0 == 0:
+                self.feed_single_frame(identifier, ea, data)
+            if len(data) > 2 and first_byte & 0xf0 == 0x10:
+                self.feed_first_frame(identifier, ea, data)
+            if len(data) > 1 and first_byte & 0xf0 == 0x20:
+                self.feed_consecutive_frame(identifier, ea, data)
+
+        def feed(self, can):
+            identifier = can.identifier
+            data = bytes(can.data)
+
+            if len(data) > 1:
+                self.try_feed(identifier, None, data)
+            if len(data) > 2:
+                ea = six.indexbytes(data, 0)
+                self.try_feed(identifier, ea, data[1:])
+
+        def count(self):
+            return len(self.ready)
+
+        def pop(self, identifier=None, ext_addr=None):
+            if identifier is not None:
+                for i in range(len(self.ready)):
+                    b = self.ready[i]
+                    identifier = b[0]
+                    ea = b[1]
+                    if identifier == identifier and ext_addr == ea:
+                        return self.ready.pop(i)
+                return None
+
+            if len(self.ready) > 0:
+                return self.ready.pop(0)
+            return None
+
+    class Closure:
+        def __init__(self):
+            self.count = 0
+            self.stop = False
+            self.max_count = 0
+            self.lst = []
+
+    @staticmethod
+    def sniff(opened_socket, count=0, store=True, timeout=None,
+              prn=None, stop_filter=None, lfilter=None, started_callback=None):
+        from scapy import plist
+        m = ISOTPSniffer.ISOTPMachine()
+        c = ISOTPSniffer.Closure()
+        c.max_count = count
+
+        def internal_prn(p):
+            m.feed(p)
+            while not c.stop and m.count() > 0:
+                rcvd = m.pop()
+                isotp_msg = ISOTP(data=rcvd[2])
+                on_pkt(isotp_msg)
+
+        def internal_stop_filter(p):
+            return c.stop
+
+        def on_pkt(p):
+            if lfilter and not lfilter(p):
+                return
+            p.sniffed_on = opened_socket
+            if store:
+                c.lst.append(p)
+            c.count += 1
+            if prn is not None:
+                r = prn(p)
+                if r is not None:
+                    print(r)
+            if stop_filter and stop_filter(p):
+                c.stop = True
+                return
+            if 0 < c.max_count <= c.count:
+                c.stop = True
+                return
+
+        opened_socket.sniff(timeout=timeout, prn=internal_prn,
+                            stop_filter=internal_stop_filter,
+                            started_callback=started_callback)
+        return plist.PacketList(c.lst, "Sniffed")
+
+
+class ISOTPSoftSocket(SuperSocket):
```

Comment:
check can_socket if filters are set to sid and did. If not, show a warning

Reply:
I have no way to check the filter on a can socket, but a warning will be emitter when a CAN frame with wrong identifier is received by the ISOTP socket for the first time

Reply:
Good solution